### PR TITLE
Replace source check with assert

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95640

cc @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -1063,7 +1063,8 @@ def wrap_to_fake_tensor_and_record(
     if type(e) in (torch.Tensor, torch.nn.Parameter) or (
         ignore_subclass and isinstance(e, torch.Tensor)
     ):
-        static_shapes, reason = tensor_always_has_static_shape(e, source, is_tensor)
+        assert source is not None
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:1066]: Assert this at the very beginning of the function and change the mypy annotation

**Review by ezyang — APPROVED** [review]
see comments

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:1066]: I am not sure this assert is valid for non tensor cases?

**Pull request closed without merging**